### ISSUE TITLE
Refactor: TensorPCA — torch-native fixes, whitening, robustness, reproducibility, incremental fit

### ISSUE THREAD

**Issue by goergen95** [goergen95/seapig#81]
## Objective
Extend/replace current TensorPCA so it is correct, robust, reproducible, memory‑efficient, and torch-native. Key goals:

- Torch-only API (torch.Tensor inputs/outputs; no numpy support required)
- Implement as a torch.nn.Module with registered buffers for persistent state
- Mutually exclusive component selection: either `n_components` OR `exp_var` (exactly one)
- `whiten` option
- Seedable randomness with torch.Generator (`random_state`) for reproducibility
- Incremental / batch-wise fitting (`partial_fit`) with deferred finalization when `exp_var` is used
- Preserve existing predict/reconstruct API (compatibility aliases) used by current implementation
- Clear tests, docstrings and maintainable implementation notes for contributors

## High-level public API

Class: TensorPCA (subclass of torch.nn.Module)

Constructor signature (torch-native):

```python
# python
class TensorPCA(torch.nn.Module):
    def __init__(
        self,
        n_components: int | None = None,
        exp_var: float | None = None,
        *,
        whiten: bool = False,
        svd_solver: str = "auto",  # one of {"auto","full","randomized","incremental"}
        batch_size: int | None = None,
        eps: float = 1e-12,
        random_state: int | torch.Generator | None = None,
        device: torch.device | None = None,
        dtype: torch.dtype = torch.float32,
    ) -> None: ...
```

Core methods:
- fit(X: torch.Tensor) -> self
- partial_fit(X: torch.Tensor) -> self
- finalize() -> self  # compute components when using partial_fit/exp_var
- fit_transform(X: torch.Tensor) -> torch.Tensor
- transform(X: torch.Tensor) -> torch.Tensor
- inverse_transform(X_proj: torch.Tensor) -> torch.Tensor
- predict(...) -> convenience wrapper (compatibility with legacy API)
- reconstruct(...) -> alias for inverse_transform / convenience wrapper

Exposed registered buffers (names and expected shapes after finalize/fit):
- components_         (Tensor, shape: [k, D])
- mean_               (Tensor, shape: [D])
- singular_values_    (Tensor, shape: [k])
- explained_variance_  (Tensor, shape: [k])
- explained_variance_ratio_ (Tensor, shape: [k])
- n_samples_seen_     (Tensor, shape: [], dtype=torch.long)

All the above are registered via `self.register_buffer(...)` so `state_dict()` round-trips.

## Design constraints and choices

1. Torch-only internals
   - Inputs and outputs are torch.Tensor. Move/cast tensors explicitly to the module device/dtype.
   - No numpy conversion; use torch.linalg primitives and torch-based randomized routines.

2. Dtype & device strategy
   - Prefer performing SVD in float64 for numerical stability: temporarily cast X to float64 on the same device, run SVD, then cast components_ back to `dtype` for storage if needed.
   - Avoid implicit CPU<->CUDA transfers. If user passes X on CUDA, SVD runs on CUDA if supported.

3. RNG and reproducibility
   - Accept `random_state` as int or torch.Generator. Internally create a torch.Generator on the module device and use it for randomized SVD draws and any stochastic steps.
   - Document that BLAS/backend/device may cause tiny numerical differences across platforms.

4. Mutual exclusivity: n_components vs exp_var
   - At the start of `fit` / `finalize`, validate that exactly one of `n_components` or `exp_var` is non-None. If not, raise ValueError.
   - If `n_components` provided: use that count (subject to max possible rank).
   - If `exp_var` provided: run SVD (or an initial compact randomized SVD) and select the smallest k such that cumulative explained_variance_ratio_ >= exp_var, then set `n_components = k` internally.
   - With `partial_fit` + `exp_var`, selection of k is deferred until `finalize()`; document this behavior and require calling `finalize()` before `transform()` when using `exp_var` in streaming mode.

5. Mean centering & incremental stats
   - Maintain `n_samples_seen_` (long tensor) and `mean_` as running mean using numerically stable incremental update.
   - For partial_fit, also keep a running second-moment accumulator (e.g., scatter matrix S = X_centered.T @ X_centered or an incremental covariance accumulator) on the module device to allow final SVD without storing all data in memory.

6. SVD solvers
   - svd_solver="full": torch.linalg.svd(X_centered, full_matrices=False)
   - svd_solver="randomized": implement randomized SVD in torch with oversampling and optional power iterations; use the internal torch.Generator for draws.
   - svd_solver="incremental": either rely on streaming SVD algorithm implemented in torch or use running scatter matrix + final SVD during finalize.
   - svd_solver="auto": heuristic based on (N, D) and batch_size.

7. Whitening
   - When `whiten=True`, scale projected components so outputs have unit variance on the training set.
   - Use stable division with `eps` to avoid division by zero: denom = singular_values_ / sqrt(n_samples_seen_ - 1) + eps.

8. Numerical robustness
   - Clip or floor tiny singular values at `eps` only when used as divisors. Preserve original singular values for explained_variance_ computations (but avoid dividing by zero).
   - Use float64 for SVD operations when feasible.

9. Partial_fit / finalize
   - partial_fit(X): update `n_samples_seen_`, `mean_`, and the running scatter/cov accumulator. Do not compute components_ until `finalize()` or an explicit call to `fit`.
   - finalize(): compute centered scatter matrix from accumulators, perform SVD on scatter (or on collected compact representation) and set all buffers (components_, singular_values_, explained_variance_, explained_variance_ratio_). If `exp_var` requested, compute k here and truncate.
   - transform() should raise a clear error if components_ is not yet computed (i.e., exp_var deferred and finalize not called).

10. Saving/loading
   - Because buffers are registered, `torch.save(pca.state_dict())` and `pca.load_state_dict(...)` work. Document expected buffer names and shapes.

## Predict / Reconstruct compatibility (legacy support)

The current implementation exposes `predict` and `reconstruct` convenience functions used by downstream code. The refactor must preserve that API and behavior as compatibility aliases:

- predict(X, return_reconstruction=False, return_error=True): convenience wrapper that computes reconstruction and/or per-sample reconstruction error. By default returns per-sample error (lower = more confident). Should accept batch inputs and preserve device/dtype.
- reconstruct(X_proj) or reconstruct(X): returns reconstructed inputs from projections or reconstructed inputs from raw X (i.e., inverse_transform(transform(X))). Offer an option to return residuals or per-sample errors.

Behavioral notes:
- `predict` / `reconstruct` use `components_`, `mean_`, and whitening settings. If components_ are not ready (partial_fit + exp_var deferred), raise informative RuntimeError instructing user to call `finalize()`.
- For backward compatibility, provide thin aliases:
  - predict -> computes transform(X) then inverse_transform and returns requested outputs
  - reconstruct -> inverse_transform
- Add tests to ensure parity with the legacy implementation (shape, dtype, device, numerical tolerances).

## Implementation notes (step details)

- Input validation: X.ndim == 2. D = X.shape[1]. If user-specified `n_components > min(N-1, D)`, cap at that rank and warn.
- Running second-moment accumulator: store S = sum_i x_i x_i^T or better store scatter in the feature space: for memory reasons store compact accumulators on device (if D is large, warn and prefer randomized solver with direct batch SVD).
- For randomized SVD: use torch.randn( device=self.device, generator=self._rng ) and implement power iterations as matrix multiplies in torch for determinism.
- For `exp_var` selection: compute explained variance from singular values: explained_variance = (S[:k]**2) / (n_samples_seen_ - 1);
  explained_variance_ratio = explained_variance / explained_variance.sum(); then choose minimal k s.t. cumulative >= exp_var.

## Tests (torch-based)

All tests should use torch tensors and torch.Generator with manual seeds. Key tests to add/adjust:
- test_pca_basic_fit_transform_torch: fit -> transform -> inverse_transform reconstruction tolerance
- test_whiten_variance_torch: whitened outputs have unit variance on training set
- test_n_components_choice_torch: n_components enforced and shapes correct
- test_exp_var_choice_torch: exp_var yields minimal k with cumulative explained variance >= exp_var
- test_mutual_exclusivity_torch: ValueError when both or neither provided
- test_random_state_reproducible_torch: repeated fits with same seed and solver produce identical components_
- test_partial_fit_equivalence_torch: partial_fit + finalize approximates fit on full batch
- test_state_dict_roundtrip_torch: save/load state_dict reproduces behavior
- test_device_and_dtype_behavior: inputs on CUDA are preserved on CUDA; SVD uses appropriate dtype promotion
- test_predict_reconstruct_torch: predict and reconstruct parity with legacy implementation (shape, dtype, device, numerical tolerances)
- test_predict_returns_error_and_reconstruction_torch: predict returns requested outputs and correct error metric
- test_predict_partial_fit_behavior_torch: predict raises informative error when components_ not ready and finalize not called

Place tests under tests/test_pca.py using pytest and torch.testing utilities.

## Tasks & priority (torch-native)

1. Add/adjust unit tests to torch-only setup (high).
2. Add input validation and mutually-exclusive `n_components` / `exp_var` checks, and raise clear errors (high).
3. Implement running stats and partial_fit + finalize logic storing accumulators as buffers (high).
4. Implement SVD solvers: full and randomized (deterministic via Generator) with dtype promotion (high).
5. Compute whitening and explained-variance bookkeeping, clip/eps protection (medium-high).
6. Implement incremental solver option or clearly document limitations and rely on finalize-based approach (medium).
7. Implement predict/reconstruct compatibility layer and add parity tests (high).
8. Add extensive docstrings, examples, and save/load docs (low-medium).
9. Microbenchmarks & solver heuristic tuning (low).

## Example usage (torch-native)

```python
# python
import torch
from seapig.scores.pca import TensorPCA

X = torch.randn(1000, 256, device="cuda")

# fixed number of components
pca = TensorPCA(n_components=64, whiten=True, svd_solver="randomized", random_state=0, device=X.device)
proj = pca.fit_transform(X)
recon = pca.inverse_transform(proj)

# backward-compatible predict/reconstruct
err = pca.predict(X)                  # default: return per-sample reconstruction error
recon2 = pca.reconstruct(pca.transform(X))

# streaming with exp_var (deferred until finalize)
pca_stream = TensorPCA(exp_var=0.95, batch_size=128, device="cpu", random_state=42)
for batch in loader:  # yields torch.Tensor (batch_N, D)
    pca_stream.partial_fit(batch)
pca_stream.finalize()  # computes components_ and decides k
proj2 = pca_stream.transform(new_queries)
```

## Acceptance criteria

- All added torch tests pass (pytest).
- Whitening behavior matches sklearn PCA whiten within numerical tolerance.
- partial_fit + finalize approximates full `fit` within tolerance when batches partition the same dataset.
- Deterministic results for fixed `random_state` on the same device/backend.
